Plugin Management: implement action status for plugin management actions

yashwin opened the pull request:
#### Proposed Changes

This PR implements the building blocks for the updated plugin details page, which includes reusing a few already existing components and rebuilding a few things. 

#### Testing Instructions

**Prerequisites**

Since this change is made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

1. Run `git checkout add/implement-plugin-action-status` and `yarn start-jetpack-cloud-p`
2. Open http://jetpack.cloud.localhost:3001/, and you'll be redirected to the /dashboard.
3. Verify action status is being shown for the actions in single-site and multi-site view for row actions and bulk actions. 

**Screenshots**

<img width="749" alt="Screenshot 2022-09-05 at 6 40 33 PM" src="https://user-images.githubusercontent.com/10586875/188467185-00cf710e-83e3-40d6-a797-1297f29f133b.png">

<img width="1614" alt="Screenshot 2022-09-05 at 6 40 45 PM" src="https://user-images.githubusercontent.com/10586875/188467199-ac12d356-6b55-499a-8828-bcb879d57f8f.png">

<img width="1587" alt="Screenshot 2022-09-05 at 6 42 17 PM" src="https://user-images.githubusercontent.com/10586875/188467212-782f188a-3ead-4b44-b128-c578302cb2f0.png">

<img width="1553" alt="Screenshot 2022-09-05 at 6 42 33 PM" src="https://user-images.githubusercontent.com/10586875/188467215-7e5c8bb0-7a21-4347-907a-2b3148d77149.png">

<img width="427" alt="Screenshot 2022-09-05 at 7 26 50 PM" src="https://user-images.githubusercontent.com/10586875/188467218-a2d15e89-f736-471f-b05d-6820552282a6.png">

<img width="499" alt="Screenshot 2022-09-05 at 7 28 35 PM" src="https://user-images.githubusercontent.com/10586875/188467221-b07d692f-6692-41a1-acea-e732a2f6653b.png">


> **_NOTE:_** Update Plugin status will be implemented in another PR

#### Pre-merge Checklist

Complete applicable items on this checklist **before** merging into the trunk. Inapplicable items can be left unchecked.

The PR author and reviewer are responsible for completing the checklist.

- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes? Will be added later - 1202518759611394-as-1202627702018877
- [x] ~~Have you tested the feature in [Simple](P9HQHe-k8-p2-p2), [Atomic](P9HQHe-jW-p2-p2), and [self-hosted Jetpack sites](PCYsg-g6b-p2-p2)?~~
- [x] Have you checked for TypeScript, React, or other console errors?
- [x] Are we memoizing when appropriate (for expensive computations)? More info in [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md) and [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors)
- [x] Have we sent any new strings [for translation](PCYsg-1vr-p2-p2) ASAP?

Related to 1202518759611394-as-1202838163671757 & 1202518759611394-as-1202663061846860

yashwin committed:
Implement action status for plugin management actions

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-row-formatter/index.tsx:
```diff
@@ -76,6 +78,22 @@ export default function PluginRowFormatter( {
 
 	const siteCount = item?.sites && Object.keys( item.sites ).length;
 
+	const inProgressStatuses = getPluginStatusesByType( state, 'inProgress' );
+	const completedStatuses = getPluginStatusesByType( state, 'completed' );
+	const errorStatuses = getPluginStatusesByType( state, 'error' );
```
I got the following TS error here. Can you see this as well?

```
Expected 1 arguments, but got 2
```

yashwin replied:
I'm unable to reproduce this. Can you please take the latest code and see if you are still getting this error?

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/index.tsx:
```diff
@@ -0,0 +1,48 @@
+		groupedStatusKeys.includes( PLUGIN_INSTALLATION_IN_PROGRESS )
+			? status === PLUGIN_INSTALLATION_IN_PROGRESS
+			: true
+	);
```
Can we try to simplify this code here, as I'm not quite sure what its main purpose is? I think if we return early for the specific case we have, the code will be a bit easier to read.

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/render-status-message.tsx:
```diff
@@ -0,0 +1,180 @@
+	};
+};
+
+export default function RenderStatusMessage( {
```
We discussed the removal of [PluginNotices](https://github.com/Automattic/wp-calypso/blob/f2e6f88ae65c8571bbac21b284c29b48ec987598/client/my-sites/plugins/notices.jsx) feedback component in the cloud, what do you think about all these informative messages we display in there? I think we should also include them here (probably, we will have to get in touch with the design team), but we can always implement these in a follow-up PR.

yashwin replied:
The messages in that component are too long to fit in here, and  I guess we are currently covering all the cases(all actions with all statuses).

yashwin replied:
What do you think, @grbicsanja?

grbicsanja replied:
I think we cover all the cases - all actions and all statuses - in a more detailed way I'd say (ie for messages like 'updated 5 plugins', we actually show which ones, and give a lot more valuable information than the global notice). I think it's safe to remove those.

In this list I see some others like:

`additionalExplanation = ( error_code ) => {
		const { translate } = this.props;

		switch ( error_code ) {
			case 'no_package':
				return translate( "Plugin doesn't exist in the plugin repo." );

			case 'resource_not_found':
				return translate( 'The site could not be reached.' );

			case 'download_failed':
				return translate( 'Download failed.' );

			case 'plugin_already_installed':
				return translate( 'Plugin is already installed.' );

			case 'incompatible_archive':
				return translate( 'Incompatible Archive. The package could not be installed.' );

			case 'empty_archive_pclzip':
				return translate( 'Empty archive.' );

			case 'disk_full_unzip_file':
				return translate( 'Could not copy files. You may have run out of disk space.' );

			case 'mkdir_failed_ziparchive':
			case 'mkdir_failed_pclzip':
				return translate( 'Could not create directory.' );

			case 'copy_failed_pclzip':
				return translate( 'Could not copy file.' );

			case 'md5_mismatch':
				return translate( "The checksum of the files don't match." );

			case 'bad_request':
				return translate( 'Invalid Data provided.' );

			case 'fs_unavailable':
				return translate( 'Could not access filesystem.' );

			case 'fs_error':
				return translate( 'Filesystem error.' );

			case 'fs_no_root_dir':
				return translate( 'Unable to locate WordPress Root directory.' );

			case 'fs_no_content_dir':
				return translate( 'Unable to locate WordPress Content directory (wp-content).' );

			case 'fs_no_plugins_dir':
				return translate( 'Unable to locate WordPress Plugin directory.' );

			case 'fs_no_folder':
				return translate( 'Unable to locate needed folder.' );

			case 'no_files':
				return translate( 'The package contains no files.' );

			case 'folder_exists':
				return translate( 'Destination folder already exists.' );

			case 'mkdir_failed':
				return translate( 'Could not create directory.' );

			case 'files_not_writable':
				return translate(
					'The update cannot be installed because we will be unable to copy some files. This is usually due to inconsistent file permissions.'
				);`

These seem to be related to installing and relevant on the plugin detail page?  We could keep these as global notices. Wdyt?

Are there any other specific messages in this list you noticed that we would lose but could be important? I could look into them and try and adjust the wording to make them shorter.

vitozev replied:
I added a new card and we can explore the opportunities here post-release: 1202518759611394-as-1202919388468349

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/render-status-message.tsx:
```diff
@@ -0,0 +1,180 @@
+	)?.[ currentGroupStatuses[ 0 ].action ]?.[ groupKey ];
+
+	return statusMessage ? (
+		<span className={ classNames( 'plugin-action-status', groupKey ) }>{ statusMessage }</span>
```
Can we replace `groupKey` class with something like `plugin-action-status-${groupKey}` to avoid having generic CSS classes like `inProgress`?

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/render-status-message.tsx:
```diff
@@ -0,0 +1,180 @@
+	const failedStatusMessage = ( message: ReactNode ) =>
+		hasOneStatus
+			? message
+			: translate( 'Failed on %(count)s site', 'Failed on %(count)s sites', translationArgs );
```
Do we need this general purpose message because we don't have enough space when there are multiple statuses?

yashwin replied:
Yes, we are using this generic failure message where there are multiple statuses; you can see the screenshot of how we show multiple statuses.

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/render-status-message.tsx:
```diff
@@ -0,0 +1,180 @@
+	hasOneStatus: boolean;
+}
+
+const getStatusMessage = (
```
Do we want to rename this to plural (messages) as it indeed returns multiple messages?

yashwin replied:
This was returning a single message only, but I have refactored the code, and now it returns multiple messages

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/render-status-message.tsx:
```diff
@@ -0,0 +1,180 @@
+}: Props ): ReactElement | null {
+	const currentGroupStatuses = groupedStatues[ groupKey ];
+
+	const statusMessage = getStatusMessage(
+		currentGroupStatuses.length,
+		hasOneStatus,
+		selectedSite
+	)?.[ currentGroupStatuses[ 0 ].action ]?.[ groupKey ];
```
I found this one difficult to understand as well. I'd recommend extracting this code to a separate function and documenting it using Docblock.

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/style.scss:
```diff
@@ -0,0 +1,37 @@
+@import '@wordpress/base-styles/_breakpoints.scss';
+@import '@wordpress/base-styles/_mixins.scss';
```
Can we replace these two with the following:

```
@import '@wordpress/base-styles/breakpoints';
@import '@wordpress/base-styles/mixins';
```
The CSS linter complains about the existing imports:

```
  1:1   ✖  Unexpected leading underscore in imported partial name            scss/at-import-no-partial-leading-underscore
  1:46  ✖  Unexpected extension ".scss" in @import                           scss/at-import-partial-extension  
```

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/style.scss:
```diff
@@ -0,0 +1,37 @@
+	}
+}
+.inProgress {
+	background: #F6F7F7;
```
Can we replace all hex codes to be all in lowercase because the CSS linter complains about it:

```
 27:14  ✖  Expected "#F6F7F7" to be "#f6f7f7"                                color-hex-case                              
 ```

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/style.scss:
```diff
@@ -0,0 +1,37 @@
+}
+.plugin-action-status {
+	position: relative;
+	top: 1px;
```
Why do we need this `top: 1px` here?

yashwin replied:
Since this element is `inline` on x-large devices, we need a little space on top of the status. I have moved this to it's breakpoint so that it is clear.

vitozev reviewed (commented):
I left some comments here and there. I also did manual testing and couldn't find any problems, but I will continue with my tests tomorrow.

yashwin committed:
Address comments and code refactoring

yashwin commented:
> I left some comments here and there. I also did manual testing and couldn't find any problems, but I will continue with my tests tomorrow.

Thanks for the quick review, @vitozev. I have addressed all your comments.

yashwin committed:
Hide PluginNotices in Jetpack Cloud

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/index.tsx:
```diff
@@ -0,0 +1,52 @@
+
+	return (
+		<div className="plugin-action-status-container">
+			{ filteredStatuses.map( ( groupKey: any ) => (
```
I think the `groupKey` has a finite number of states and we can explicitly list them here?

vitozev commented on client/my-sites/plugins/plugin-management-v2/plugin-action-status/render-status-message.tsx:
```diff
@@ -0,0 +1,40 @@
+		>
+			{ currentStatusMessage }
+		</span>
+	) : null;
```
It's a nitpick and definitely a personal preference, but returning early would be more readable.

```
	if ( ! currentStatusMessage ) {
		return null;
	}

	return (
		<span
			className={ classNames( 'plugin-action-status', `plugin-action-status-${ currentStatus }` ) }
		>
			{ currentStatusMessage }
		</span>
	);
```

vitozev reviewed (approved):
LGTM! I tested all different actions (without update) and they are working as expected, and inline statuses are displayed correctly. Nice job!

I left two minor comments, which are definitely not blocker for moving this PR forward.

yashwin committed:
Return empty state early

grbicsanja commented:
@yashwin Could we add a 2px vertical padding to the text in the notices?
<img width="1298" alt="Screenshot 2022-09-06 at 12 00 26" src="https://user-images.githubusercontent.com/90194580/188606993-6cafc262-e455-4842-bbc8-b5ae9d6b3bcc.png">

yashwin committed:
Increase vertical padding of the action status badge

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7513321

Thank you @yashwin for including a screenshot in the description! This is really helpful for our translators.

grbicsanja commented:
Looks great!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7513321).
